overlord/snapstate/snapstate.go

```diff
@@ -1396,13 +1394,84 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []
 	return updateManyFiltered(ctx, st, names, revOpts, userID, nil, flags, "")
 }
 
-// EnforceSnaps installs/updates/removes snaps reported by validationErrorToSolve.
-// validationSets is the list of sets passed by the user and it's used in the
-// final stage to update validation-sets tracking in the state.
-// userID is used for store auth.
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+// ResolveEnforcementError installs and updates snaps reported by validationErrorToSolve.
+func ResolveEnforcementError(ctx context.Context, st *state.State, vsMap map[string]*asserts.ValidationSet, valErr *snapasserts.ValidationSetsValidationError, userID int) ([]*state.TaskSet, []string, error) {
```

Comment:
This maybe should be called  ResolveValidationSetsEnforcementError, the error should probably come before the map in the arguments given the name of the function